File: api/core/pii_scrub.py
"""PII 遮蔽共用工具（CR-0166 R1-8）。

原實作在 core/observability.py（OTel span 用）；本模組昇格為共用 util，供
audit payload 遮蔽（入 hash 鏈前）、log filter 等接管道使用。observability
改 import 此處，避免兩份 regex drift。

兩個變體：
  - scrub_text：全遮蔽（含地址啟發式）——log / OTel span 用。
  - scrub_audit_value：高置信樣式（email/電話/身分證/LINE uid，不含地址啟發式）——
    audit payload 用，避免地址啟發式誤殺稽核證據力（CR-0166 §8-D10 裁決）。
"""

from __future__ import annotations

import hashlib
import re

# 順序有意義：LINE uid 先於 token（U 開頭 33 字元）；電話先於地址（門牌數字）。
_LINE_UID_RE = re.compile(r"\bU[0-9a-f]{32}\b")
_EMAIL_RE = re.compile(r"[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}")
# 台灣手機（09xxxxxxxx / +8869xxxxxxxx）與市話（0x-xxxxxxxx）
_PHONE_RE = re.compile(
    r"(?:\+886[-\s]?9\d{2}|09\d{2})[-\s]?\d{3}[-\s]?\d{3}"
    r"|\b0\d{1,2}-\d{6,8}\b"
)
# 台灣身分證：1 大寫英文字母 + 1/2 開頭 + 8 位數字（CR-0166 R1-8 新增）
_NATIONAL_ID_RE = re.compile(r"\b[A-Z][12]\d{8}\b")
# 台灣地址啟發式（保守，寧漏勿誤殺）——僅 scrub_text 用，audit 不用
_ADDR_RE = re.compile(
    r"\S{1,6}[縣市]\S{0,12}?[區鄉鎮市]?\S{0,20}?(?:路|街|大道|巷|弄)[\S]{0,12}?號?"
)
_TOKEN_PARAM_RE = re.compile(r"((?:access_|refresh_)?token=)[^&\s]+")


def _hash_line_uid(m: re.Match) -> str:
    return "U#" + hashlib.sha256(m.group(0).encode()).hexdigest()[:12]


def _scrub_common(value: str) -> str:
    """LINE uid / token / 身分證 / email / 電話（不含地址）。"""
    value = _LINE_UID_RE.sub(_hash_line_uid, value)
    value = _TOKEN_PARAM_RE.sub(r"\1[TOKEN]", value)
    value = _NATIONAL_ID_RE.sub("[ID]", value)
    value = _EMAIL_RE.sub("[EMAIL]", value)
    value = _PHONE_RE.sub("[PHONE]", value)
    return value


def scrub_text(value: str) -> str:
    """全遮蔽（含地址啟發式）——log / OTel span 用。非 PII 內容原樣保留。"""
    if not isinstance(value, str):
        return value
    value = _scrub_common(value)
    value = _ADDR_RE.sub("[ADDR]", value)
    return value


def scrub_audit_value(value):
    """audit payload 值遮蔽：高置信樣式，不含地址啟發式（保稽核證據力）。
    非字串原樣回傳。"""
    if not isinstance(value, str):
        return value
    return _scrub_common(value)


def scrub_audit_payload(payload):
    """遞迴遮蔽 dict/list 內所有字串值（audit payload 專用，入 hash 前呼叫）。"""
    if isinstance(payload, dict):
        return {k: scrub_audit_payload(v) for k, v in payload.items()}
    if isinstance(payload, list):
        return [scrub_audit_payload(v) for v in payload]
    return scrub_audit_value(payload)

File: web/platform-console/src/observability/piiScrub.ts
/**
 * OTel span 屬性 PII 深度遮蔽（25_Monitoring §3 硬性要求）——四站共同遞延項銷案。
 *
 * 範式移植 api/core/pii_scrub.py 與 agent/lockcore/observability.py：regex 順序
 * 有意義（LINE uid 先於 token；電話先於地址）。掛法＝自訂 SpanProcessor 於 onEnd
 * 就地遮蔽字串屬性，registerOTel({ spanProcessors: [scrub, "auto"] })——陣列順序
 * 即執行順序，'auto'（預設 Batch+OTLP 匯出）在後 → 出站前必先過遮蔽。
 * 遮蔽失敗絕不 throw（可觀測性不可癱瘓服務；該 span 屬性原樣照出）。
 * 本檔四站位元組級一致（brand-portal / tech-portal / landing / platform-console）。
 */
import type { ReadableSpan, Span, SpanProcessor } from "@opentelemetry/sdk-trace-base";

const LINE_UID_RE = /\bU[0-9a-f]{32}\b/g;
const TOKEN_PARAM_RE = /((?:access_|refresh_)?token=)[^&\s]+/g;
const EMAIL_RE = /[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}/g;
const PHONE_RE =
  /(?:\+886[-\s]?9\d{2}|09\d{2})[-\s]?\d{3}[-\s]?\d{3}|\b0\d{1,2}-\d{6,8}\b/g;
const ADDR_RE =
  /\S{1,6}[縣市]\S{0,12}?[區鄉鎮市]?\S{0,20}?(?:路|街|大道|巷|弄)[\S]{0,12}?號?/g;

function hashLineUid(uid: string): string {
  try {
    // 64-bit FNV-1a 僅產生穩定 pseudonymous fingerprint；LINE UID 本身為
    // 128-bit 隨機識別碼。純 JS 實作可同時在 Node/Edge/client fallback 編譯，
    // 避免 instrumentation 將 node:crypto 誤帶進 Next.js browser bundle。
    let fingerprint = BigInt("0xcbf29ce484222325");
    for (const char of uid) {
      fingerprint ^= BigInt(char.charCodeAt(0));
      fingerprint = BigInt.asUintN(
        64,
        fingerprint * BigInt("0x100000001b3"),
      );
    }
    return "U#" + fingerprint.toString(16).padStart(16, "0").slice(0, 12);
  } catch {
    return "[LINE_UID]";
  }
}

/** 單一字串遮蔽：LINE uid（雜湊）→ token 參數 → email → 電話 → 地址。 */
export function scrubText(value: string): string {
  let v = value.replace(LINE_UID_RE, hashLineUid);
  v = v.replace(TOKEN_PARAM_RE, "$1[TOKEN]");
  v = v.replace(EMAIL_RE, "[EMAIL]");
  v = v.replace(PHONE_RE, "[PHONE]");
  v = v.replace(ADDR_RE, "[ADDR]");
  return v;
}

/** 出站前遮蔽 processor：只動 string / string[] 屬性，數值與 bool 原樣。 */
export class PIIScrubSpanProcessor implements SpanProcessor {
  onStart(_span: Span): void {
    // 進場不動作——屬性在 span 生命週期內可能續增，統一於 onEnd 遮蔽
  }

  onEnd(span: ReadableSpan): void {
    try {
      const attrs = span.attributes as Record<string, unknown> | undefined;
      if (!attrs) return;
      for (const key of Object.keys(attrs)) {
        const v = attrs[key];
        if (typeof v === "string") {
          const s = scrubText(v);
          if (s !== v) attrs[key] = s;
        } else if (Array.isArray(v) && v.some((i) => typeof i === "string")) {
          attrs[key] = v.map((i) => (typeof i === "string" ? scrubText(i) : i));
        }
      }
    } catch {
      // 遮蔽失敗絕不影響 span 匯出（該 span 屬性原樣照出）
    }
  }

  forceFlush(): Promise<void> {
    return Promise.resolve();
  }

  shutdown(): Promise<void> {
    return Promise.resolve();
  }
}
